### PR TITLE
feature: Next Auth Adapter

### PR DESCRIPTION
**Pull request by siddhsuresh**
### 🚧Testing
*Please give feedback here if not merged yet or in a discord thread in help or in contributor general if you know to fix the bug*

**‼️ This PR has been published to npm to allow testing**

Install Using tag `@next--nextauth`

#### Current API
```ts
import { api } from "src/blitz-server"
import GithubProvider from "next-auth/providers/github"
import GoogleProvider from "next-auth/providers/google"
import { NextAuthAdapter } from "@blitzjs/auth/next-auth"
import db, { User } from "db"
import { Role } from "types"

// Has to be defined separately for `profile` to be correctly typed below
const providers = [
  GithubProvider({
    clientId: process.env.GITHUB_CLIENT_ID as string,
    clientSecret: process.env.GITHUB_CLIENT_SECRET as string,
  }),
  GoogleProvider({
    clientId: process.env.GOOGLE_CLIENT_ID as string,
    clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
  }),
]

export default api(
  NextAuthAdapter({
    successRedirectUrl: "/",
    errorRedirectUrl: "/error",
    providers,
    callback: async (user, account, profile, session) => {
      if(account?.provider === "google") {
      }
      console.log("USER SIDE PROFILE_DATA", { user, account, profile })
      let newUser: User
      try {
        newUser = await db.user.findFirstOrThrow({ where: { name: { equals: user.name } } })
      } catch (e) {
        newUser = await db.user.create({
          data: {
            email: user.email as string,
            name: user.name as string,
            role: "USER",
          },
        })
      }
      const publicData = {
        userId: newUser.id,
        role: newUser.role as Role,
        source: "github",
      }
      await session.$create(publicData)
      return { redirectUrl: "/" }
    },
  })
)

```

#### New wrapper for `next.config.js`
```ts
const { withNextAuthAdapter } = require("@blitzjs/auth/next-auth")
const { withBlitz } = require("@blitzjs/next")

/**
 * @type {import('@blitzjs/next').BlitzConfig}
 **/
const config = {
  reactStrictMode: true,
}

module.exports = withBlitz(withNextAuthAdapter(config))
```

TODO:
- [x] Add redirect export option

## Feature Checklist

- [x] Changeset added (run `pnpm changeset` in the root directory)
- [x] Documentation added/updated (submit PR to [blitzjs.com repo `main` branch](https://github.com/blitz-js/blitzjs.com))

### CODE REVIEW

**Comment by siddhsuresh** [comment]
!preview next-auth

File: packages/blitz-auth/build.config.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import {BuildConfig} from "unbuild"
 
 const config: BuildConfig = {
   entries: ["./src/index-browser", "./src/index-server"],
-  externals: ["index-browser.cjs", "index-browser.mjs", "react"],
+  externals: ["index-browser.cjs", "index-browser.mjs", "react", "next-auth", "next-auth/jwt"],
```
Comment (flybayer) [inline packages/blitz-auth/build.config.ts:5]: add as an optional peer dependency instead of here, otherwise pnpm may not work right

**Review by flybayer — CHANGES_REQUESTED** [review]
sweet!! Worked perfectly. Few design changes to make:

- [ ] change `[...auth].ts` to `[...nextauth].ts` for better readability
- [ ] rename `signin` to `login`
- [ ] change config structure so that each provider can have it's own callback

instead of
```ts
  providers: [
    GithubProvider({}),
  ],
  callback: async (user, account, profile, session) => {}
```
do like we have for passport strategy
```ts
  providers: [
    {
      provider: GithubProvider({}),
      callback: async (user, account, profile, session) => {})
     }
  ],
```

- [ ] along with the above change, I think we can auto extract the type from the provider so that callback is properly typed ([see nextauth src](https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/providers/github.ts#L59))
- [ ] add a script for updating next-auth version in our repo

**Comment by flybayer** [comment]
and then also go ahead and write docs

File: packages/blitz-auth/build.config.ts (hunk 1 above)
Comment (siddhsuresh) [reply packages/blitz-auth/build.config.ts:5]: okay I will do that

Comment (flybayer) [reply packages/blitz-auth/build.config.ts:5]: still todo

File: packages/blitz-auth/src/server/adapters/next-auth/adapter.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      Array.isArray(req.query.auth),
+      "req.query.auth must be an array. Page must be named [...auth].ts/js.",
+    )
+    if (!req.query.auth?.length) {
+      return res.status(404).end()
+    }
+    const action = req.query.auth[0] as AuthAction
+    if (!action || !["signin", "callback"].includes(action)) {
+      return res.status(404).end()
+    }
```
[the remark this answers is not in the document]
Comment (flybayer) [reply packages/blitz-auth/src/server/adapters/next-auth/adapter.ts:47]: still todo, I think?

File: packages/blitz-auth/src/server/adapters/next-auth/internals/core/lib/cookie.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
+import type {CookiesOptions} from "next-auth"
```
Comment (flybayer) [inline packages/blitz-auth/src/server/adapters/next-auth/internals/core/lib/cookie.ts:1]: we need **all** of these files that are next-auth/internals?

File: packages/blitz-auth/src/server/adapters/next-auth/types.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import {SessionContext} from "../../../index-server"
+
+export type BlitzNextAuthOptions = AuthOptions & {
+  successRedirectUrl: string
+  errorRedirectUrl: string
+  secureProxy?: boolean
+  callback: (
+    user: User,
+    account: any,
+    profile: Profile,
```
Comment (flybayer) [inline packages/blitz-auth/src/server/adapters/next-auth/types.ts:13]: We want `profile` type to extract the actual profile type from the given `providers`. Do you need help with that?

File: packages/blitz-auth/src/server/adapters/next-auth/internals/core/lib/cookie.ts (hunk 3 above)
Comment (siddhsuresh) [reply packages/blitz-auth/src/server/adapters/next-auth/internals/core/lib/cookie.ts:1]: From my experimentation with the next auth patch. We need to import only the 3 function the oAuthCallback for the callback the getAuthorizedUrl for the login and the init to generate the request and the options to the required way next auth requires.

These other files I have copied are all dependencies of those functions.

File: packages/blitz-auth/src/server/adapters/next-auth/types.ts (hunk 4 above)
Comment (siddhsuresh) [reply packages/blitz-auth/src/server/adapters/next-auth/types.ts:13]: Yeah, This is where I spend most time in. I tried both the approach, the common callback and the individual callback.
I get mostly stuck with the circular or implicit invocation type errors.
I think I could use some help/insights.

I will post my attempts in Discord and could continue this there

File: packages/blitz-auth/src/server/adapters/next-auth/adapter.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (siddhsuresh) [reply packages/blitz-auth/src/server/adapters/next-auth/adapter.ts:47]: Yes I will push the changes soon (mostly by tonight)

**Review by flybayer — APPROVED** [review]

**Pull request merged**